src/parse/core/routine.hpp

```diff
@@ -0,0 +1,152 @@
+#pragma once
+#include <cstring>
+#include <array>
+#include <utility>
+#include "state.hpp"
+
+namespace docgen {
+namespace parse {
+namespace core {
+
+// Tags indicate possible routines 
+enum class Routine {
+    READ = 0,
+    IGNORE_WS,
+    PROCESS,
+    NUM_ROUTINES
+};
+
+// Forward-declaration of a routine.
+// routine will parse the chunk of text starting at begin
+// until it reaches end, given current state, and update the json object.
+// If a state changes, routine will modify state.
+// When a state changes or the chunk has been fully read, the routine will finish.
+// It will always return the next routine to execute.
+// routine will perform differently depending on routine_tag.
+
+template <Routine routine_tag>
+struct routine
+{
+    template <class Cache>
+    static Routine run(Cache& cache, const char*& begin, const char* end);
+};
+
+namespace details {
+
+template <class Cache>
+inline Routine process_slash(Cache& cache, const char*& begin, const char* end)
```

Comment:
It's a nice helper function that makes the logic for the whole `read routine` simpler. We do not need a separate function like this for every symbol or keyword. You are getting confused because `//` and `/*` are just 2 letters. Later when we process, we will need a similar helper function for when we hit `[` and we are ready to parse "the next tag". Once it hits `]`, it checks if the word it just read is a valid tag. Note that I'm not making a helper function for every letter after `[`, or even a helper function for every keyword after `[`.

Reply:
Also I took away the state associated with this function. Notice that it's in `namespace details`. There's only a `read` routine that calls this. If you won't even let me write a helper function idk what you want me to do buddy.

Reply:
I would propose we do something more generalized—actually take advantage of C++'s features. Whether it's `/*` or `[...]`, these are all the same problem—we need a structure which will look for these continuous strings of characters as we go. You're going to end up doing an impossible amount of work if we have to hard-code every possible character sequence as we see here.

Reply:
Ah, how about creating a function where given a char* begin, char* end, and char* substring, we extract everything from begin to end until we hit substring and return that? I think that's what you mean. That's v nice!

Reply:
I guess the "opposite" of this would be the `read` circle in our diagram. We read strings of characters until we hit a substring either `//` or `/*` or `[`.

Reply:
Yes, I'm actually working on something for this. The common thread here is that we're pretty much always looking for an opening string, a closing string, and looking for a routine/action to take on the data in between. I think if we can think of a structure to encapsulate all this, we solve quite a few of our current problems and future problems.

Reply:
Awesome I totally agree with this approach